### PR TITLE
Fix codeAction

### PR DESCRIPTION
**Pull request by mattn**
Language Server Specification says

>The arguments are typically specified when a command is returned from the server to the client. Example requests that return a command are textDocument/codeAction or textDocument/codeLens.

This fix codeAction in jdtls.

### CODE REVIEW

**Comment by thomasfaingnaert** [comment]
Shouldn't we rather store the `arguments` we get from `textDocument/codeAction` and pass these to `workspace/executeCommand`?

File: autoload/lsp/ui/vim.vim [hunk 1]
```diff
@@ -606,16 +606,23 @@ endfunction
 "   server - string
 "   comand_or_code_action - Command | CodeAction
 function! s:execute_command_or_code_action(server, command_or_code_action) abort
-    if has_key(a:command_or_code_action, 'command') && type(a:command_or_code_action['command']) == type('')
-        let l:command = a:command_or_code_action
-        call s:execute_command(a:server, l:command)
+    let l:code_action = a:command_or_code_action
+
+    if has_key(l:code_action, 'arguments') && type(l:code_action['arguments']) == type([])
+        for l:action in l:code_action['arguments']
```
Comment (thomasfaingnaert) [inline autoload/lsp/ui/vim.vim]: My understanding is that `arguments` can be any data, and is only used at the server side. I don't think you should be interpreting this at the client?

**Comment by thomasfaingnaert** [comment]
Eclipse JDT LS indeed doesn't work:
```
Fri 13 Sep 2019 18:54:59 CEST:["<---",1,"eclipse",{"response":{"id":9,"jsonrpc":"2.0","error":{"code":-32601,"message":"No delegateCommandHandler for java.apply.workspaceEdit"}},"request":{"id":9,"jsonrpc":"2.0","method":"workspace/executeCommand","params":{"arguments":[{"changes":{"file:///home/thomas/Program.java":[{"range":{"end":{"character":0,"line":3},"start":{"character":0,"line":0}},"newText":""}]}}],"command":"java.apply.workspaceEdit"}}}]
```
but isn't this a bug in the language server?

**Comment by thomasfaingnaert** [comment]
I think I know what is happening here. The LSP spec states:

> # Command
>
> Represents a reference to a command. Provides a title which will be used to represent a command in the UI. Commands are identified by a string identifier. The recommended way to handle commands is to implement their execution on the server side if the client and server provides the corresponding capabilities. Alternatively the tool extension code could handle the command. The protocol currently doesn’t specify a set of well-known commands.

So, the way it should work is:
1. A client sends `textDocument/codeAction` to the server.
2. The servers returns a list of `Command`s or `CodeAction`s to the client.
3. The ciient shows these in the UI, and the user picks one.
4. The selected `Command` is then sent to the server via a `workspace/executeCommand`
5. The server then determines what this command does, and sends (for example) `workspace/applyEdit` back to the client.
6. The `workspace/applyEdit` is then interpreted by the client, and the text edits are applied.

This ensures that the code actions are useful in as many clients as possible.
It seems that Eclipse JDT LS breaks this scheme by having the VSCode extension intercept this and do the textEdits instead?

In any case, interpreting the `arguments` of the Command shouldn't be done in the client. The content is not defined by the spec, precisely because it's only used in the server.

**Comment by mattn** [comment]
So do you mean that this is a bug of jdtls?

**Comment by prabirshrestha** [comment]
@mattn @thomasfaingnaert any update on this?

**Comment by thomasfaingnaert** [comment]
@prabirshrestha Am I right in assuming https://github.com/prabirshrestha/vim-lsp/pull/505#issuecomment-531319355 is the way codeActions are supposed to work? If so, I think we should avoid adding server-specific functionality to vim-lsp by interpreting the `arguments` field. What are your thoughts on this?

**Comment by prabirshrestha** [comment]
We shouldn't be interpreting it.

Wondering how does this works in vscode currently? Do they show a menu to us because they have registered a delegate handler?

This message seems interesting.  `No delegateCommandHandler for java.apply.workspaceEdit`.

I'm wondering if `vim-lsp-settings` should register the handler and do what they want.

**Comment by thomasfaingnaert** [comment]
My guess is that the VSCode plugin intercepts the request and applies the edit, instead of the language server responding with a `workspace/applyEdit`?

**Comment by prabirshrestha** [comment]
@mattn this is not a bug on the server. It is just that it is custom.

I think what we should instead do is allow users to customize the callback of this. Then we can move the fix to `vim-lsp-settings`.

Here is an example where vscode-java is doing a lot more than just edit.

https://github.com/redhat-developer/vscode-java/blob/7f005ee59b67d1879a21abc377d50407daa9ed3b/src/refactorAction.ts#L33-L94

These are custom extensions that lsp protocol allows and is a bug on the server nor it is a standard so all languages implements it unfortunately.

**Comment by mattn** [comment]
Thanks. And I think `LspCodeAction` should accept arguments. For example, vscode have `codeActionsOnSave` to specify codeAction to run when file is saved.

**Comment by mattn** [comment]
What style of customization for the callback?

**Comment by prabirshrestha** [comment]
Should be part of registering the server. https://github.com/prabirshrestha/vim-lsp/blob/master/doc/vim-lsp.txt#L634-L641. Might be have `hooks: {}` so we can extend it to add more in the future? 

If a server exposes it we should completely ignore vim-lsp apply edit which means we might have to expose a method to the user so they can manually call apply text editor or workspace edits.

**Comment by mattn** [comment]
Is it for only codeAction? This problem occur since vim-lsp send request (applyEdit) that is given from executeCommandProvider eventhough the command should execute for my-self. Do you think the hook should be located at?

**Comment by prabirshrestha** [comment]
My thought was this:

```vim

function! s:on_code_action(server_name, request, response) abort
   " here the user can do what ever they want with request and response.
   " lsp#utils#workspace_edit#apply_workspace_edit is already exposed so we should be able to call it
endfunction

au User lsp_setup call lsp#register_server({
        \ 'name': 'pyls',
        \ 'cmd': {server_info->['pyls']},
        \ 'whitelist': ['python'],
        \ 'hooks': { 'codeaction': function('s:on_code_action') }
        \ })
```

Instead of executing `s:handle_code_action` would need to run `hooks.codeaction`. Since the user now has raw access to request and response in `s:on_code_action`, they can do anything.

We should come up with a better name for the hooks instead of just `codeaction` since I would also like it in the future to be extended with `precodeactions` so that we can even change the requests.

This will only work for `codeaction` but we could also use hooks for edits. `'hooks': { 'applyEdit': function('s:on_apply_edit') }`.

Would this work? Let me know if I'm missing something here.

**Comment by mattn** [comment]
I'm still wondering why the hooks solve this issue. Does some servers expect a response to the server in response to the codeAction response, while others don't?

**Comment by mattn** [comment]
Once I close this issue since https://github.com/prabirshrestha/vim-lsp/pull/663 was merged. I'll probably create new PR for hooks.

**Comment by prabirshrestha** [comment]
Sounds good.

**Pull request closed without merging**